### PR TITLE
feat: add no-missing-space-atx rule

### LINKED ISSUE #370: New Rule: no-space-after-hash-on-atx-style-heading
### Rule details

This rule will warn when spaces are missing after the hash characters in an atx style heading:

### What type of rule is this?

Warns about a potential problem

### Example code

```markdown
#Heading 1

##Heading 2
```

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

Prior At: [Rule](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md018---no-space-after-hash-on-atx-style-heading)

### PR DESCRIPTION
**Pull request by SwetaTanwar**
#### Prerequisites checklist

- [ ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?
This PR adds a new rule `no-missing-space-atx` to ensure there is a space after hash on ATX style headings in Markdown

#### What changes did you make? (Give an overview)
Added the `no-missing-space-atx` rule, along with documentation and tests.

#### Related Issues
fixes #370

### CODE REVIEW

**Review by lumirlumir — CHANGES_REQUESTED** [review]
Could you please take a look at the CI failures? 

I think there might be some edge cases that were missed.

File: docs/rules/no-missing-space-atx.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# no-missing-space-atx
+
+This rule warns when spaces are missing after the hash characters in an ATX style heading.
+
+## Rule Details
+
+In Markdown, headings can be created using ATX style (using hash characters at the beginning of the line) or Setext style (using underlining with equals or hyphens).
```
Comment (snitin315) [inline docs/rules/no-missing-space-atx.md:7]:
```suggestion
In Markdown, headings can be created using ATX style (using hash (`#`) characters at the beginning of the line) or Setext style (using underlining with equals (`=`) or hyphens (`-`).
```

File: docs/rules/no-missing-space-atx.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Heading 1
+## Heading 2
+### Heading 3
+```
+
+Lines that contain hash characters but are not headings (not at the beginning of the line) will not be flagged:
+
+```md
+This is a paragraph with a #hashtag, not a heading.
+```
```
Comment (snitin315) [inline docs/rules/no-missing-space-atx.md:33]: Let's also mention the following valid cases for links & images:

```
[#123](link.com)

![#alttext][link.png]
```

File: tests/rules/no-missing-space-atx.test.js [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+//------------------------------------------------------------------------------
+
+const validHeadings = [
+	// 1. ATX headings with proper spacing
+	"# Heading 1",
+	"## Heading 2",
+	"### Heading 3",
+	"#### Heading 4",
+	"##### Heading 5",
+	"###### Heading 6",
```
Comment (snitin315) [inline tests/rules/no-missing-space-atx.test.js:37]: html headings should also be valid.
```suggestion
	"###### Heading 6",
	"<h1>Heading 1<h1>",
	"<h2>Heading 2<h2>",
	"<h3>Heading 3<h3>",
	"<h4>Heading 4<h4>",
	"<h5>Heading 5<h5>",
	"<h6>Heading 6<h6>",
	"<h1>#valid heading</h1>"
```

Comment (snitin315) [reply tests/rules/no-missing-space-atx.test.js:37]: Quotes should also be valid:
```suggestion
	"###### Heading 6",
	`'#something'`
	`"#something"`
```

Comment (SwetaTanwar) [reply tests/rules/no-missing-space-atx.test.js:37]: > Maybe we can just traverse the paragrph node and check for the first child

@snitin315 This was a great approach—thanks for this suggestion! It really simplified the code and helped avoid unnecessary node traversals. 🙌

**Comment by nzakas** [comment]
How about `no-missing-atx-header-space`? I think reads a bit clearer.

**Comment by lumirlumir** [comment]
Would it be better to use `no-missing-atx-heading-space`?

CommonMark Spec and Mdast use `heading`.

- https://spec.commonmark.org/0.31.2/#atx-headings

image
image

**Comment by lumirlumir** [comment]
@nzakas Thanks, I'll keep that in mind!

**Comment by SwetaTanwar** [comment]
@nzakas @lumirlumir  Renamed the package from `no-missing-atx-header-space` to `no-missing-atx-heading-space`

**Review by nzakas — DISMISSED** [review]
LGTM. Would like @lumirlumir to review before merging.

**Review by snitin315 — APPROVED** [review]
LGTM, thanks!

**Review by lumirlumir — CHANGES_REQUESTED** [review]
Thanks for your hard work on this!

I’ve left a few comments about some missing parts.

**Comment by SwetaTanwar** [comment]
@lumirlumir I have addressed your feedback, PTAL

**Review by lumirlumir — CHANGES_REQUESTED** [review]

File: src/rules/no-missing-atx-heading-space.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * @typedef {import("../types.ts").MarkdownRuleDefinition<{ RuleOptions: []; }>}
+ * NoMissingAtxHeadingSpaceRuleDefinition
+ */
+
+//-----------------------------------------------------------------------------
+// Rule Definition
+//-----------------------------------------------------------------------------
+
+const HEADING_PATTERN = /^(#{1,6})(?:[^#\s])/u;
```
[the remark this answers is not in the document]
Comment (SwetaTanwar) [reply src/rules/no-missing-atx-heading-space.js:19]: As per the naming conventions, the standard is to use UPPERCASE for global variables.

[the remark this answers is not in the document]
Comment (SwetaTanwar) [reply src/rules/no-missing-atx-heading-space.js:19]: Added comments for helpers

File: src/rules/no-missing-atx-heading-space.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		},
+	},
+
+	create(context) {
+		return {
+			paragraph(node) {
+				if (node.children && node.children.length > 0) {
+					const firstTextChild = node.children.find(
+						child => child.type === "text",
+					);
```
[the remark this answers is not in the document]
Comment (SwetaTanwar) [reply src/rules/no-missing-atx-heading-space.js:51]: I have updated the regex to handle `#*hi*`

File: tests/rules/no-missing-atx-heading-space.test.js [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({
+	plugins: {
+		markdown,
+	},
+	language: "markdown/commonmark",
```
Comment (lumirlumir) [inline tests/rules/no-missing-atx-heading-space.test.js:23]:
```suggestion
	language: "markdown/gfm",
```

Could you change it to `gfm`? If you're trying to use `strikethrough` in your test case, you'll need to set the language to `gfm` since it's a GFM-only syntax.

File: tests/rules/no-missing-atx-heading-space.test.js [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+	// 6. Emphasis handling
+	{
+		code: "#*hi*",
+		output: "# *hi*",
+		errors: [
+			{
+				messageId: "missingSpace",
+				column: 1,
+				line: 1,
+				endLine: 1,
```
Comment (lumirlumir) [inline tests/rules/no-missing-atx-heading-space.test.js:371]:
```suggestion
				endLine: 6,
```

It should be the length of `#*hi*`, since other errors follow the same error location format.

The same applies to the other Emphasis handling test cases.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

File: docs/rules/no-missing-atx-heading-space.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+In Markdown, headings can be created using ATX style (using hash (`#`) characters at the beginning of the line) or Setext style (using underlining with equals (`=`) or hyphens (`-`)).
+
+For ATX style headings, a space should be used after the hash characters to improve readability and ensure proper rendering across various Markdown parsers.
+
+This rule is automatically fixable by the `--fix` command line option.
+
+Examples of **incorrect** code for this rule:
+
+```md
+#Heading 1
```
Comment (snitin315) [inline docs/rules/no-missing-atx-heading-space.md:16]:
```suggestion
```md
<!-- eslint markdown/no-missing-atx-heading-space: "error" -->

#Heading 1
```

File: docs/rules/no-missing-atx-heading-space.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+##Heading 2
+###Heading 3
+```
+
+Examples of **correct** code for this rule:
+
+```md
+# Heading 1
```
Comment (snitin315) [inline docs/rules/no-missing-atx-heading-space.md:24]:
```suggestion
```md
<!-- eslint markdown/no-missing-atx-heading-space: "error" -->

# Heading 1
```

File: src/rules/no-missing-atx-heading-space.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (SwetaTanwar) [reply src/rules/no-missing-atx-heading-space.js:51]: As per markdown-lint they are just showing error on the intial column not the whole end of line.
Attaching screenshot for your reference

Screenshot 2025-05-20 at 9 34 19 PM

Screenshot 2025-05-20 at 9 34 42 PM

[the remark this answers is not in the document]
Comment (snitin315) [reply src/rules/no-missing-atx-heading-space.js:51]: Hmm, yeah, reporting the location like markdownlint does makes sense. I think it would also be easier to implement.

@lumirlumir WDYT?

```md
<!-- eslint markdown/no-missing-atx-heading: "error" -->

#test
^^

  ##test
  ^^^

######test
^^^^^^^

##*test*
^^^

#__test__
^^
```

File: tests/rules/no-missing-atx-heading-space.test.js [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+				endColumn: 1,
+			},
+		],
+	},
```
Comment (snitin315) [inline tests/rules/no-missing-atx-heading-space.test.js:417]: Let's add a few test cases for quoted headings like:

```md

> #Quoted heading

> ##Quoted heading

- Item
  > #Quoted heading in list
```

File: tests/rules/no-missing-atx-heading-space.test.js (hunk 7 above)
Comment (SwetaTanwar) [reply tests/rules/no-missing-atx-heading-space.test.js:388]: Fixed the endLine as per https://github.com/eslint/markdown/pull/371#discussion_r2098403409

**Comment by lumirlumir** [comment]
@SwetaTanwar

I apologize if my review comments came across as a bit aggressive. My English is still a work in progress, so there may have been some unintended tone. I’ll make sure to use kinder wording next time.

I’ll review it in a few days — thank you in advance for your patience!

**Comment by SwetaTanwar** [comment]
@lumirlumir, No worries at all — I didn’t find the tone off in any way. Really appreciate your comments, they helped catch some important cases I missed. 

Looking forward to your next round of feedback when you get a chance! 😊

File: src/rules/no-missing-atx-heading-space.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (lumirlumir) [reply src/rules/no-missing-atx-heading-space.js:51]: It looks good to me 👍

**Review by lumirlumir — APPROVED** [review]
LGTM! Thanks for your hard work and patience on this PR.

Would like to ask @snitin315 and @nzakas to re-verify before merging.

**Review by snitin315 — APPROVED** [review]
LGTM, thank you for contributing ⭐
